### PR TITLE
Generate a call of a base class constructor if the default constructor does not exist.

### LINKED ISSUE #11254: Generate a call of a base class constructor if the default constructor does not exist.
# Context
```C#
class A
{
    public A(int a, bool b) { }
}

class B : A
{
    public B() : base(default, default) { }
}
```

Base class **A** does not have default constructor, that's why derived classes must have explicit constructor and call base class constructor.

The question: is it possible having a `INamedTypeSymbol` object (corresponds to A in an example) to determine if it has default constructor or not and if not how to determine the list of parameters for explicit constructor (From example: `A(int a, bool b)`).

Having a constructor (object of `IMethodSymbol`), we'd like to generate compliable code `public B() : base(default, default) { }`

### PR DESCRIPTION
**Pull request by andriipatsula**
Fixes: https://github.com/dotnet/arcade/issues/11254

### CODE REVIEW

**Review by smasher164 — APPROVED** [review]
LGTM

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs [hunk 1]
```diff
@@ -43,5 +64,24 @@ public static SyntaxNode DeclarationExt(this SyntaxGenerator syntaxGenerator, IS
...
+            ConstructorInitializerSyntax constructorInitializer = SyntaxFactory.ConstructorInitializer(SyntaxKind.BaseConstructorInitializer);
+
+            IMethodSymbol baseConstructor = baseTypeConstructors.OrderBy(c => c.Parameters.Count()).First();
+            foreach (IParameterSymbol parameter in baseConstructor.Parameters)
+            {
+                IdentifierNameSyntax identifier;
+                if (parameter.Type.IsValueType)
+                    identifier = SyntaxFactory.IdentifierName("default");
+                else
+                    identifier = SyntaxFactory.IdentifierName("default!");
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:79]: Is that what we did in the legacy GenAPI?

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs [hunk 2]
```diff
@@ -43,5 +64,24 @@ public static SyntaxNode DeclarationExt(this SyntaxGenerator syntaxGenerator, IS
                 throw new ArgumentException(ex.Message, symbol.ToDisplayString());
             }
         }
+
+        private static ConstructorInitializerSyntax GenerateBaseConstructorInitializer(ImmutableArray<IMethodSymbol> baseTypeConstructors)
+        {
+            ConstructorInitializerSyntax constructorInitializer = SyntaxFactory.ConstructorInitializer(SyntaxKind.BaseConstructorInitializer);
+
+            IMethodSymbol baseConstructor = baseTypeConstructors.OrderBy(c => c.Parameters.Count()).First();
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:72]: The legacy GenAPI handled this case as well: https://github.com/dotnet/arcade/blob/608f3c7aed68dec1314d75601c6f143747e78cc6/src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Methods.cs#L444-L445

Do we want to port that over here as well?

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 1 above)
Comment (ViktorHofer) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:79]: This comment suggests that there should be a better way to figure out if a parameter type accepts null or not: https://github.com/dotnet/arcade/blob/608f3c7aed68dec1314d75601c6f143747e78cc6/src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Methods.cs#L456-L463

cc @eerhardt

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs [hunk 3]
```diff
@@ -2,8 +2,14 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Data.Common;
+using System.Linq;
+using System.Security.AccessControl;
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:9]: This using is Interesting. What API from Security.AccessControl is used?

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs [hunk 4]
```diff
@@ -33,6 +39,21 @@ public static SyntaxNode DeclarationExt(this SyntaxGenerator syntaxGenerator, IS
...
             }
 
+            if (symbol.Kind == SymbolKind.Method)
+            {
+                IMethodSymbol method = (IMethodSymbol)symbol;
+                if (method.MethodKind == MethodKind.Constructor)
+                {
+                    INamedTypeSymbol? baseType = method.ContainingType.BaseType;
+                    // If the base type does not have default constructor.
+                    if (baseType != null && !baseType.Constructors.IsEmpty && baseType.Constructors.Where(c => c.Parameters.IsEmpty).Count() == 0)
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:49]: When using LINQ, we usually prefer `!Any()` over `Count() == 0` but in this case, an All operator might even be more readable.

```suggestion
                    if (baseType != null && !baseType.Constructors.IsEmpty && baseType.Constructors.All(c => !c.Parameters.IsEmpty))
```

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 1 above)
Comment (ViktorHofer) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:79]: image

This might be a the right algorithm to identify when to use `default` vs `default!`.

File: src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs [hunk 5]
```diff
@@ -1116,5 +1116,42 @@ public partial class C : B
...
+        [Fact]
+        public void TestBaseTypeConstructorWithObsoleteAttribute()
+        {
+            RunTest(original: """
+                    namespace Foo
+                    {
+                        public class B
+                        {
+                            public B(int p1, string p2) { }
+                            [System.Obsolete("Constructor is deprecated.", true)]
```
Comment (ViktorHofer) [inline src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:1129]: Can you please add two more test case for when the obsoleteattribute is applied but without the error: true parameter and when the obsolete message is null?

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 2 above)
Comment (ViktorHofer) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs]: Unsure but can we just mimic the existing legacy GenAPI behavior? cc @ericstj

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs [hunk 6]
```diff
@@ -46,10 +43,17 @@ public static SyntaxNode DeclarationExt(this SyntaxGenerator syntaxGenerator, IS
...
                     INamedTypeSymbol? baseType = method.ContainingType.BaseType;
                     // If the base type does not have default constructor.
-                    if (baseType != null && !baseType.Constructors.IsEmpty && baseType.Constructors.Where(c => c.Parameters.IsEmpty).Count() == 0)
+                    if (baseType != null && !baseType.Constructors.IsEmpty && baseType.Constructors.All(c => !c.Parameters.IsEmpty))
                     {
-                        ConstructorDeclarationSyntax declaration = (ConstructorDeclarationSyntax)syntaxGenerator.Declaration(method);
-                        return declaration.WithInitializer(GenerateBaseConstructorInitializer(baseType.Constructors));
+                        IOrderedEnumerable<IMethodSymbol> baseTypeConstructors = baseType.Constructors
+                            .Where(c => !c.GetAttributes().Any(a => a.IsObsoleteWithUsageTreatedAsCompilationError()))
+                            .OrderBy(c => c.Parameters.Count());
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:50]: An analyzers inside VS suggests to use Length instead of Count():

```suggestion
                            .OrderBy(c => c.Parameters.Length);
```

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs [hunk 7]
```diff
@@ -46,10 +43,17 @@ public static SyntaxNode DeclarationExt(this SyntaxGenerator syntaxGenerator, IS
                 {
                     INamedTypeSymbol? baseType = method.ContainingType.BaseType;
                     // If the base type does not have default constructor.
-                    if (baseType != null && !baseType.Constructors.IsEmpty && baseType.Constructors.Where(c => c.Parameters.IsEmpty).Count() == 0)
+                    if (baseType != null && !baseType.Constructors.IsEmpty && baseType.Constructors.All(c => !c.Parameters.IsEmpty))
                     {
-                        ConstructorDeclarationSyntax declaration = (ConstructorDeclarationSyntax)syntaxGenerator.Declaration(method);
-                        return declaration.WithInitializer(GenerateBaseConstructorInitializer(baseType.Constructors));
+                        IOrderedEnumerable<IMethodSymbol> baseTypeConstructors = baseType.Constructors
+                            .Where(c => !c.GetAttributes().Any(a => a.IsObsoleteWithUsageTreatedAsCompilationError()))
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:49]: same here: Use All instead of !Any for readability:

```suggestion
                            .Where(c => c.GetAttributes().All(a => !a.IsObsoleteWithUsageTreatedAsCompilationError()))
```

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs [hunk 8]
```diff
@@ -65,15 +69,14 @@ public static SyntaxNode DeclarationExt(this SyntaxGenerator syntaxGenerator, IS
...
         {
             ConstructorInitializerSyntax constructorInitializer = SyntaxFactory.ConstructorInitializer(SyntaxKind.BaseConstructorInitializer);
 
-            IMethodSymbol baseConstructor = baseTypeConstructors.OrderBy(c => c.Parameters.Count()).First();
-            foreach (IParameterSymbol parameter in baseConstructor.Parameters)
+            foreach (IParameterSymbol parameter in baseTypeConstructor.Parameters)
             {
                 IdentifierNameSyntax identifier;
-                if (parameter.Type.IsValueType)
+                if (parameter.Type.IsValueType || parameter.NullableAnnotation == NullableAnnotation.Annotated)
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:79]: Please add a code comment to explain what this conditions checks as it isn't obvious.

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs [hunk 9]
```diff
@@ -46,10 +43,17 @@ public static SyntaxNode DeclarationExt(this SyntaxGenerator syntaxGenerator, IS
                 {
                     INamedTypeSymbol? baseType = method.ContainingType.BaseType;
                     // If the base type does not have default constructor.
-                    if (baseType != null && !baseType.Constructors.IsEmpty && baseType.Constructors.Where(c => c.Parameters.IsEmpty).Count() == 0)
+                    if (baseType != null && !baseType.Constructors.IsEmpty && baseType.Constructors.All(c => !c.Parameters.IsEmpty))
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:46]: We are missing a filter check here to only include constructors that are publicly accessible (i.e. filter out sealed ones): https://github.com/dotnet/arcade/blob/608f3c7aed68dec1314d75601c6f143747e78cc6/src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Methods.cs#L444-L445

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 2 above)
Comment (ericstj) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs]: I'd start by saying that GenAPI doesn't need to handle a case that's impossible to represent in the product.  It also shouldn't go out of its way to support something that's extremely unlikely (eg: we Obsolete(true) an API that is in active use by some up-stack component).

Worst case, if we somehow had a case of this scenario, then I would expect that we'd need to workaround it by having an internal constructor on the base type. Manual exclusion could also allow a user to exclude the upstack constructor if they needed to redefine it to call a non-default internal constructor.  I would recommend that GenAPI omit the base call to an obsolete(true) constructor.  This is what the [old GenAPI did](https://github.com/dotnet/arcade/blob/608f3c7aed68dec1314d75601c6f143747e78cc6/src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Methods.cs#L444-L445).  

TL;DR
1. Omit all calls to base obsolete(true) constructors.
2. Developers can author an internal constructor on the base type manually (and a call it, if it's not default).

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 7 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:49]: My expectation is that `Any` returns false when the first element found as a result in best case you don't need to traverse the whole list. On the other hand `All` will always traverse the list.

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 6 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs]: updated

File: src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs (hunk 5 above)
Comment (andriipatsula) [reply src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:1129]: Added 2 more unit tests

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 9 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:46]: members are filter out before we generate SyntaxNode out of ISymbol. https://github.com/dotnet/sdk/blob/bfbcfe92e8df98f04dc5e1a8835b0270176f4b67/src/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs#L121

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 8 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:80]: added a comment

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 7 above)
Comment (ViktorHofer) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:49]: > On the other hand All will always traverse the list.

That would be a very inefficient implementation of All. It returns as soon as it finds an element that doesn't match: https://source.dot.net/#System.Linq/System/Linq/AnyAll.cs,62

`All(a => !a.IsObsoleteWithUsageTreatedAsCompilationError())` is the just the more readable statement in comparison to `!Any(a => a.IsObsoleteWithUsageTreatedAsCompilationError())`.

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 9 above)
Comment (ViktorHofer) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:46]: The filtering makes sure that we only traverse symbols that are included. Here we are looking at the base type constructors of the included constructor. Nothing guarantees at that point that the base type constructors are included.

You can test that with a type that has a base type with an internal ctor. This should result in a compilation error because the internal ctor of the base type isn't emitted (assuming that the `IncludeVisibleOutsideOfAssembly` setting is false).

Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs]: Thanks, updated.

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 7 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs]: updated

**Review by ViktorHofer — APPROVED** [review]

**Review by ViktorHofer — CHANGES_REQUESTED** [review]
The filter includes effectively sealed constructors, doesn't it? https://github.com/dotnet/sdk/blob/a08e0a55cb7d31e873dd3a10b6ea03eca1fc7146/src/GenAPI/Microsoft.DotNet.GenAPI/GenAPIApp.cs#L95

We need to find a non effectively sealed ctor.

Also, please add a test with an effectively sealed ctor to guard this path. Effectively sealed means i.e. when the ctor is protected (= publicly exposed) but the type is sealed (so nobody can derive from it).

**Review by ViktorHofer — APPROVED** [review]
Chatted offline. Effectively sealed ctors don't matter in this case as other types can't inherit from such types anyway.

**Review by ViktorHofer — APPROVED** [review]
